### PR TITLE
ENH: Infer best datetime format from a sample

### LINKED ISSUE #52508: ENH: Try both dayfirst and monthfirst in pd.to_datetime
### Feature Type

- [ ] Adding new functionality to pandas

- [X] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

If you run `pd.to_datetime` on the following Series:
```
    "11-12-2029",
    "02-12-2012",
    "11-09-2012",
    "13-02-2000",
    "10-11-2001"
```
pandas (>= 2.0) will infer the datetime format from the first non-missing example (%m%d%Y), try to apply this type to all the series, fail on `13-02-2000`, and raise an error (before version 2.0, this would silently create a mixed type). I wish pandas could infer the right format from such a series, where only one format works for all rows. 

## Pseudo code
If using `dayfirst=True` and `dayfirst=False` *don't* give the same format for `guess_datetime_format` on the first non missing example (i.e both works):
Try both formats on the Series (probably on a random subset for speed).
If one works for all rows, return this format.
If both work, trust the dayfirst parameter (and maybe raise a warning).
If none work and `error="raise"`, raise an error. If `errors = "coerce"` or `errors="ignore"`, one could either trust the `dayfirst` parameter, or see which of `dayfirst` value leads to the smallest number of non-parsed values.

## Implementation
Change function `_guess_datetime_format_for_array` (in `pandas.core.tools.datetimes`) so that it tries both `dayfirst=True` and `dayfirst=False` on the first non-null example. In the same function, if both options give a different format, try `array_strptime` with both format on a random subset of the array (100?) with strict error, and check that one of the tries doesn't fail.

### Alternative Solutions

I don't know.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by LeoGrin**
- [x] closes #52508 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

## Summary
``pd.to_datetime`` now tries to infer the datetime format of each string by considering
a random sample (instead of the first non-null sample),
and tries to find the format which work for most strings. If several
formats work as well, the one which matches the ``dayfirst`` parameter is returned. If
``format="mixed"``, pandas does the same thing, then tries the second best format on the
strings which failed to parse with the first best format, and so on (instead of parsing each row
independently) (#52508).

*Previous behavior*:

```
    In [1]: pd.to_datetime(["01-02-2012", "01-03-2012", "30-01-2012"])
    Out[1]:
    ValueError: time data "30-01-2012" doesn't match format "%m-%d-%Y", at position 2. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.
    In [2]: pd.to_datetime(["01-02-2012", "01-03-2012", "30-01-2012"], errors="coerce")
    Out[2]:
    DatetimeIndex(['2012-01-02', '2012-01-03', 'NaT'], dtype='datetime64[ns]', freq=None)
    In [3]: pd.to_datetime(["01-02-2012", "01-03-2012", "30-01-2012"], format="mixed")
    Out[3]:
    DatetimeIndex(['2012-01-02', '2012-01-03', '2012-01-30'], dtype='datetime64[ns]', freq=None)
```
*New behavior*:

```
    In [1]: pd.to_datetime(["01-02-2012", "01-03-2012", "30-01-2012"])
    Out[1]:
    UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False was specified.
    Pass `dayfirst=True` or specify a format to silence this warning.
    DatetimeIndex(['2012-02-01', '2012-03-01', '2012-01-30'], dtype='datetime64[ns]',
    freq=None)
    In [2]: pd.to_datetime(["01-02-2012", "01-03-2012", "30-01-2012"], errors="coerce")
    Out[2]:
    UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False was specified. Pass `dayfirst=True` or specify a format to silence this warning.
    DatetimeIndex(['2012-02-01', '2012-03-01', '2012-01-30'], dtype='datetime64[ns]', freq=None)
    In [3]: pd.to_datetime(["01-02-2012", "01-03-2012", "30-01-2012"], format="mixed")
    Out[3]:
    DatetimeIndex(['2012-02-01', '2012-03-01', '2012-01-30'], dtype='datetime64[ns]', freq=None)
```

## Design questions:
- Do we bypass the `dayfirst` parameter too often? Right now, we always prefer the format which matches the most strings, and only consider `dayfirst` if several formats match as many strings (and raise a warning if we contradict `dayfirst`). For `error="raise"` it makes sense (we only select a format if it matches all non-null values), but I'm wondering if it's a problem for `errors="coerce"` or `errors="ignore"`: in this case, if `dayfirst = True` but `%m-%d%-%Y` matches more values than `%d-%m%-%Y`, we select `%m-%d%-%Y` (and raise a warning). I think it's okay but it may be somewhat confusing for users. Some ideas otherwise:
    - Have `dayfirst=None` (instead of `False`) as the default in `pd.to_datetime`, and respect `dayfirst` if it's provided.
    - Have some threshold of match percentage to be able to overwrite `dayfirst`: for instance  if `dayfirst = True`, `%m-%d%-%Y` needs to match >10% more values than `%d-%m%-%Y` to be chosen.

### CODE REVIEW

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for working on this

couple of initial comments:
- this looks very complicated, is a simpler solution possible?
- this should be deterministic, and shouldn't depend on the result of `np.random` - can we use, say, the first 10 non-null elements? Or 10 equally spaced elements?

**Comment by LeoGrin** [comment]
Thanks for the quick reply!
> * this looks very complicated, is a simpler solution possible?
Probably, I'll try to make it simpler.
> * this should be deterministic, and shouldn't depend on the result of `np.random` - can we use, say, the first 10 non-null elements? Or 10 equally spaced elements?
Will fix

**Comment by LeoGrin** [comment]
@MarcoGorelli this is ready for review (don't know if I should tag or just wait)
The failing build seems to be related to #52853, not related to the PR.

File: doc/source/whatsnew/v0.19.0.rst [hunk 1]
```diff
@@ -765,6 +765,7 @@ Previously if ``.to_datetime()`` encountered mixed integers/floats and strings,
 This will now convert integers/floats with the default unit of ``ns``.
 
 .. ipython:: python
+   :okwarning:
 
    pd.to_datetime([1, "foo"], errors="coerce")
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v0.19.0.rst:770]: why is this needed?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for updating!

this still look quite complicated - this is going to have be maintained long-term by multiple people and so I'm a bit hesitant to add a lot of code. Is there a simpler solution which would still improve datetime format inference?

**Comment by MarcoGorelli** [comment]
> maybe for now we could directly fall back to dateutil when format="mixed" is provided?

yeah `format='mixed'` means that no format inference should take place

**Comment by MarcoGorelli** [comment]
Thanks for your PR

I'm afraid this introduces too much complexity, sorry, it's going to be too hard to maintain, there's already very few people comfortable with this part of the codebase 

If you can find a simpler solution which improves format inference without greatly increasing the maintenance burden, then we can take. For example, just trying 5 elements and taking a majority vote would probably be an improvement and still be quite simple

**Pull request closed without merging**